railgun/pkg/config/config.go

```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
 	}
 	return client.New(conf, client.Options{})
 }
+
+func (c *Config) ConfigKongService(ctx context.Context) error {
+	// deployment configuration
+	adminApiService := os.Getenv("CONTROLLER_KONG_ADMIN_PUBLISH_SERVICE")
+	if adminApiService == "" {
+		// parameter configuration
+		if len(c.KongAdminURL) > 0 && c.KongAdminURL != "http://localhost:8001" {
+			return nil
+		}
+		return fmt.Errorf(" kong admin api url %s should not be localhost.", c.KongAdminURL)
```

Comment:
question: what led us to make this assertion that the kong admin API URL should not be localhost? It seems to me that we should be allowing this, and making this change I worry will in fact break things? :thinking:

Reply:
The function is a bit confusing, since it manipulates the user config struct (which we want to get away from) as a side effect. I think the error text is inaccurate--it looks like the effect here is instead to _only_ allow `c.KongAdminURL == "http://localhost:8001"` if it is set along with `CONTROLLER_KONG_ADMIN_PUBLISH_SERVICE`. If so, `ConfigKongService()` exits _without_ modifying the config struct, i.e. if `c.KongAdminURL` was set to `http://localhost:8001`, it remains so. If those conditions aren't true, `ConfigKongService()` proceeds, composes a URL from the Service, and updates the config struct with that URL.

I don't think this behavior is correct (it doesn't handle common cases like a non-standard admin API port or HTTPS admin API listen on localhost, and relies on extracting configuration directly from the environment instead of using the `config` package--something we should only use for test environment special cases), but can't really evaluate that without a clearer understanding of the design goals here, hence my [original question](https://github.com/Kong/kubernetes-ingress-controller/pull/1487#issuecomment-874862259) asking why we'd add this in the first place.

Reply:
Here is the clarification: 
assuming admin api url configuration "http://localhost:8001" is valid and generic configuration, if deployments does not specify configuration (admin api url service name and namespace), leave as it is line 216. 

Otherwise, specified name/namespace will be applied to retrieve the exact service IP/hostname and port, which is used to compose the in cluster URL i.e  http://172.18.0.240:8001

The whole point is to read information from in cluster configuration (no hardcode).